sdks/python/apache_beam/ml/gcp/recommendations_ai.py

```diff
@@ -0,0 +1,638 @@
+      catalog_name='my-catalog')
+    """
+  def __init__(
+      self,
+      project: str = None,
+      retry: Retry = None,
+      timeout: float = 120,
+      metadata: Sequence[Tuple[str, str]] = None,
+      catalog_name: str = "default_catalog"):
+    """Initializes a :class:`CreateCatalogItem` transform.
+
+        Args:
+            project (str): Optional. GCP project name in which the catalog
+              data will be imported.
+            retry (google.api_core.retry.Retry): Optional. Designation of what
+              errors, if any, should be retried.
+            timeout (float): Optional. The amount of time, in seconds, to wait
+              for the request to complete.
+            metadata (Sequence[Tuple[str, str]]): Optional. Strings which
+              should be sent along with the request as metadata.
+            catalog_name (str): Optional. Name of the catalog.
+              Default: 'default_catalog'
+        """
+    self.project = project
+    self.retry = retry
+    self.timeout = timeout
+    self.metadata = metadata
+    self.catalog_name = catalog_name
+
+  def expand(self, pcoll):
+    if self.project is None:
+      self.project = pcoll.pipeline.options.view_as(GoogleCloudOptions).project
+    if self.project is None:
+      raise ValueError(
+          """GCP project name needs to be specified in "project" pipeline
+            option""")
+    return pcoll | ParDo(
+        _CreateCatalogItemFn(
+            self.project,
+            self.retry,
+            self.timeout,
+            self.metadata,
+            self.catalog_name))
+
+
+class _CreateCatalogItemFn(DoFn):
+  FAILED_CATALOG_ITEMS = "failed_catalog_items"
+
+  def __init__(
+      self,
+      project: str = None,
+      retry: Retry = None,
+      timeout: float = 120,
+      metadata: Sequence[Tuple[str, str]] = None,
+      catalog_name: str = None):
+    self._client = None
+    self.retry = retry
+    self.timeout = timeout
+    self.metadata = metadata
+    self.parent = f"projects/{project}/locations/global/catalogs/{catalog_name}"
+    self.counter = Metrics.counter(self.__class__, "api_calls")
+
+  def setup(self):
+    if self._client is None:
+      self._client = get_recommendation_catalog_client()
+
+  def process(self, element):
+    catalog_item = recommendationengine.CatalogItem(element)
+    request = recommendationengine.CreateCatalogItemRequest(
+        parent=self.parent, catalog_item=catalog_item)
+
+    try:
+      created_catalog_item = self._client.create_catalog_item(
+          request=request,
+          retry=self.retry,
+          timeout=self.timeout,
+          metadata=self.metadata)
+
+      self.counter.inc()
+      yield recommendationengine.CatalogItem.to_dict(created_catalog_item)
+    except Exception:
+      yield pvalue.TaggedOutput(
+          self.FAILED_CATALOG_ITEMS,
+          recommendationengine.CatalogItem.to_dict(catalog_item))
+
+
+class _BatchItems(DoFn):
```

Comment:
Like in Java, Python also has a GroupIntoBatches transform that also has autosharding:

- https://beam.apache.org/documentation/transforms/python/aggregation/groupintobatches/
- https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.util.html#apache_beam.transforms.util.GroupIntoBatches